TreeViewer: auto expansion for elements with single children

Wittmaxi opened the pull request:
As implementation of #1063, This PR addresses the need to automatically expand paths which consist of chained singular nodes, eg:

```
com.wittmaxi.plugin
 └─src
    └─org.foo.com
       └─Bar.java
```
It introduces two new API-Methods into TreeViewer:

-setAutoExpandTrivialPathsLevel
-getAutoExpandTrivialPathsLevel

The PR also uses the new features in the Project Explorer, thus resolving #1063

Wittmaxi commented:
I still need to address a few points as part of this PR

- [x] When double-clicking on an item, the TreeListener-hook is not called
- [x] There are no Unit-Tests yet
- [x] The behaviour should be off by default
- [x] What should be the default value for `expandTrivialPathsLevel`?

merks commented:
For auto expand there are only getAutoExpandLevel/setAutoExpandLevel with 0 and -1 having special meaning so it's seems a little strange/confusing four methods for this new feature...

Wittmaxi commented:
Good catch, I changed my proposal to be more consistent with the existing behaviour @merks

merks commented:
I think this will be a cool feature!

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1229,6 +1278,16 @@ public int getAutoExpandLevel() {
 	}
 
+	/**
+	 * @return 0 for disabled, <code> ALL_LEVELS </code> for infinite expansion or
```
Shouldn't the disabled state be represented by `NO_EXPAND` rather than `0`?

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2466,6 +2525,28 @@ public void setAutoExpandLevel(int level) {
 	}
 
+	/**
+	 * Sets the level of auto expand on Trivial Paths. 0 means that trivial paths
```
See above: no auto-expansion should be reflected by `NO_EXPAND` rather than `0`.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -140,6 +155,40 @@ protected AbstractTreeViewer() {
+	protected Item[] expandAndCreateChildren(Widget w) {
+		setExpanded((Item) w, true);
+		createChildren(w);
+		return getChildren(w);
```
This seems to reimplement the essential expansion functionality as already implemented in `setExpandedState`. Is there a necessity for this additional method?

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2597,11 +2678,14 @@ public void setExpandedState(Object elementOrTreePath, boolean expanded) {
+		if ((item instanceof Item)) {
 			if (expanded) {
-				createChildren(item);
+				if (expandAndCreateChildren(item).length == 1) {
```
I would expect the additional extension to happen after the "ordinary" extension has happended completely, i.e., after creating childen and setting the expand state has happened. First performing the recusive extension and then setting the expand state does not look to be the right order to me.

Wittmaxi replied:
I have changed the order in which both operations happen, I suspect that this does not correctly reflect what you actually intended by your comment. If that is the case, can you please specify what exactly you have in mind?

HeikoKlare replied:
Recursive expansion should happen by completely expanding an items before starting the expansion of a child. The state change of an auto expansion should be equal to manually extending levels one-by-one, i.e., before expanding a child, the parent expansion should be finished in terms of creating all its children, settings it expanded state (and maybe further state changes I am not aware of). Currently, the logic is quite convoluted, as this method sets the expanded state and then calls `expandAndCreateChildren`, which sets the state again.

Wittmaxi replied:
```java
if (item instanceof Item) {
	setExpanded((Item) item, expanded);
	if (expanded) {
		if (expandAndCreateChildren(item).length == 1) {
			expandPathOfSingleChildren(getChildren(item)[0], getAutoExpandOnSingleChildLevel());
		}
	}
}
```

What currently happens:
1) set the expanded state (calls the winapi, doesn't add the children)
**if the expanded state is expanded: <call expandAndCreateChildren>**
2) set the expanded state again *(doesn't do anything new)*
3) create the children
4) return the amount of children
5) check: does my parent only have one child? If so, start the recursive routine that expands trivial paths and keeps track of how many levels it has recursed

Steps 1-4 expand the node and do not yet add any behaviour. Step 5 is the actual expansion of paths of singular children - just as you had described in your comment, I first perform the existing behaviour and then simulatie "clicking" through the children to consecutively open them.
From what I understood in your comment, your issue boils down to 1) and 2) being duplicates of eachother. A rather obvious solution would be to check for the expanded state so that I only set it once

```java
if (!expanded) {
  setExpanded((Item) item, false);
} else {
  if (expandAndCreateChildren(item).length == 1) {
    expandPathOfSingleChildren(getChildren(item)[0], getAutoExpandOnSingleChildLevel());
  }
}
```

I am a bit hesitant about making this change as it introduces additional control flow and complexity. However, `setExpanded` performs a call to winapi (... and probably to cocoa and gtk too), hence I can understand why we would want to limit it's use.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/TreeViewer.java:
```diff
@@ -246,6 +247,20 @@ protected void hookControl(Control control) {
 		super.hookControl(control);
 		Tree treeControl = (Tree) control;
 
+		treeControl.addTreeListener(new TreeListener() {
```
The listener should be added and removed when setting the variable for the number of levels to expand in order to only have the listener attached when auto-expansion is enabled. In addition, the current implementation would lead to implementations of `AbstractTreeViewer` that do not extend `TreeViewer` to not automatically expand even if the according values have been set. This means, the chage on `AbstractTreeViewer` currently is API breaking, as all extensions of that class need to be adapted to fulfill the changed method contracts.

HeikoKlare commented on bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/navigator/resources/ProjectExplorer.java:
```diff
@@ -165,6 +166,7 @@ public void createPartControl(Composite aParent) {
 		if (this.userFilters.stream().anyMatch(UserFilter::isEnabled)) {
 			getCommonViewer().refresh();
 		}
+		getCommonViewer().setAutoExpandTrivialPathsLevel(AbstractTreeViewer.ALL_LEVELS);
```
Enabling auto-expansion in a specific usage context should be done in a separate commit to providing the basic functionality.

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -151,6 +151,54 @@ public void testExpandToLevel() {
+		fTreeViewer.setAutoExpandTrivialPathsLevel(1);
+		fTreeViewer.setExpandedState(trivialPathRoot, true);
+
+		assertTrue("The expanded widget child is expanded", fTreeViewer.getExpandedState(trivialPathRoot));
```
Applies to all assertion messages: the message describe the failure state, not the success state, i.e., the messages have to be "inverted".

HeikoKlare reviewed (changes requested):
I like the proposed change in general, still I have two major remarks (and some fine-grained ones made as code comments):

**Functionality:**
The changes to `AbstractTreeViewer` are currently API breaking: the new methods for setting the auto expansion (such as `setAutoExpandTrivialPathsLevel`) claim that when setting the expansion level, the instance will perform auto-expansion. The current implementation does, however, not guarantee that but relies on implementations of that class (such as `TreeViewer`) to provide the according behavior. I would propose to change the implementation such that `AbstractTreeViewer` already provides the claimed functionality by changing the according methods handling changes of the expansion state of a node and adding a listener that reacts to expansion state changes (like now done in the `TreeViewer`.

**Naming:**
I am not in favor of the current naming, as I don't think that there is a common understanding of what "trivial paths" are. This requires every method related to the functionality to explain what is meant. I would suggest to change the naming to something more self-explaning. This could be something like `autoExpandOnSingleChild` or the like (better proposals are welcome).

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -101,8 +110,23 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
 	/**
-	 * Indicates if filters should be checked to determine expandability of
-	 * a tree node.
+	 * How many levels to autoexpand in a trivial path. 0 will disable the feature,
```
```suggestion
	 * How many levels to autoexpand in a trivial path. {@code NO_EXPAND} will disable the feature,
```

HeikoKlare replied:
Please replace the term "trivial path" in Javadoc.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2457,6 +2513,58 @@ public void setAutoExpandLevel(int level) {
+	 * auto-expansion of trivial paths
+	 * </p>
+	 *
+	 * @param level 0 for disabled, <code> ALL_LEVELS </code> for infinite expansion
```
```suggestion
	 * @param level {@code NO_EXTEND} for disabled, <code> ALL_LEVELS </code> for infinite expansion
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -140,6 +164,27 @@ protected AbstractTreeViewer() {
+	 * @param maxDepth how deep to recursively expand a trivial path
+	 * @since 3.31
+	 */
+	protected void expandTrivialPath(Widget w, int maxDepth) {
```
Replease consistently replace the termin "trivial path".

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2457,6 +2513,58 @@ public void setAutoExpandLevel(int level) {
 	}
 
+	/**
+	 * Sets the level of auto expand on Trivial Paths. <code> NO_EXPAND </code>
```
Please replace the term "trivial path" in Javadoc.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2573,26 +2681,40 @@ public int hashCode(Object element) {
 		Widget item = internalExpand(elementOrTreePath, false);
-		if (item instanceof Item) {
+
+		if ((item instanceof Item)) {
```
Remove duplicated brackets

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/TreeViewer.java:
```diff
@@ -86,6 +86,7 @@ public class TreeViewer extends AbstractTreeViewer {
 
 	private boolean contentProviderIsTreeBased;
 
+
```
Please remove whiteline change.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2457,6 +2513,58 @@ public void setAutoExpandLevel(int level) {
+	}
+
+	private void removeAutoExpandOnSingleChildListener() {
+		if (autoExpandOnSingleChildListener != null)
```
Please format single line condition block as an ordinary block (with brackets).

HeikoKlare reviewed (commented):
I like that you were able to move functionality completely into `AbstractTreeViewer`. That makes my previous concerns with respect to API compatibility obsolete.

Please consider my unanswered comment from last review.

HeikoKlare reviewed (commented):
I have taken a deeper look into the overall design of the expand mechanism in the `AbstractTreeViewer`. I would like to make some proposals to improve this PR by reusing existing functionality and the insights behind.

1. `internalExpandToLevel` already has a sophisticated implementation for expanding a certain amount of levels, including some performance improvements. I would propose to let the recursive expansion when only having a single child use this functionality by generalizing it to some kind of `internalConditionalExpandToLevel` that additionally accepts a `shouldExpand` function (Widget -> boolean) that is used to calculate whether the next recursion of expansion shall be performed. Then the existing `internalExpandToLevel` method can use this method with a function always returning true, while the expand on single child functionality will pass a condition based on the number of childs of the widget.
2. The `expandToLevel(...)` methods disable the redraw of the tree viewer while expanding (to avoid flickering). That may not be a severe problem with recursive expansion when only having one child, as this should be quite fast, still the functionality is already there and I would thus propose to reuse it. You could, for example, encapsulate the transaction that ensures proper disablement and re-enablement of redraw into a functional that can be used by both the existing `expandToLevel` as well as the `expandOnSingleChild`. It could look something like this:
```java
private void excuteWithRedrawDisabled(Runnable toExecute) {
	Control control = getControl();
	try {
		control.setRedraw(false);
		toExecute.run();
	} finally {
		control.setRedraw(true);
	}
}
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1819,41 +1850,68 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
+	 * @since 3.32
+	 */
+	protected void internalConditionalExpandToLevel(Widget widget, int level,
+			Function<Widget, Boolean> shouldFunctionRecurseDown) {
```
`shouldFunctionRecurseDown` is an implementation-specific, technical name (as it indicates that this method performs recursion).
Please replace it by a user-oriented name, such as `shouldWidgetExpand`.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1819,41 +1850,68 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
 			Object data = widget.getData();
-			if (widget instanceof Item && data != null
-					&& !isExpandable((Item) widget, null, data)) {
+			if (widget instanceof Item && data != null && !isExpandable((Item) widget, null, data)) {
```
Please only reformat lines that you actually change.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1819,41 +1850,68 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
+	 *               levels of the tree
+	 */
+	protected void internalExpandToLevel(Widget widget, int level) {
+		internalConditionalExpandToLevel(widget, level, w -> Boolean.valueOf(true));
```
```suggestion
		internalConditionalExpandToLevel(widget, level, w -> Boolean.TRUE);
```

HeikoKlare reviewed (commented):
I thought that the refactoring of the temporary redraw disablement would be rather local and thus proposed to do it in this PR or extract it to another that is merged before. Since a reasonable refactoring will probably be a bit larger (see #1191) and should not prevent this PR from being merged, So feel free to change the code back to the original sequence of manual setRedraw(false) and setRedraw(true), which will then be refactored together with all other analogous places in the code in an according refactoring. Sorry for the back and forth, but I was not aware that the issue is a more general one when proposing the refactoring.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -102,8 +112,24 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
 	/**
-	 * Indicates if filters should be checked to determine expandability of
-	 * a tree node.
+	 * How many levels to autoexpand in a path of consecutive single children. 0
```
```suggestion
	 * How many levels to autoexpand in a path of consecutive single children. {code NO_EXPAND}
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -102,8 +112,24 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
-	 * Indicates if filters should be checked to determine expandability of
-	 * a tree node.
+	 * How many levels to autoexpand in a path of consecutive single children. 0
+	 * will disable the feature, while a negative value (<code>ALL_LEVELS</code>)
```
You should never use the value of a constant for documentation, as that value may change (that's one of the reason why it was defined as a constant).
```suggestion
	 * will disable the feature, while <code>ALL_LEVELS</code>
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1817,22 +1854,31 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
-	 * level.
+	 * Recursively, conditionally expands the subtree rooted at the given widget to
+	 * the given level. Takes a predicate as input which takes in a Widget and
+	 * returns a boolean value.
```
Please make a user-oriented description rather than a technical one, e.g., "Takes the {@code shouldChildrenExpend} predicate that defines for a given Widget if it shall be expanded."

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1817,22 +1854,31 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
+	 *                             expanded further. The predicate-function takes in
+	 *                             a Widget-type and returns a Boolean value, true
+	 *                             indicating that the Widget should be expanded
+	 *                             further
```
Please use a more simple and less technical description, such as the one I proposed above. In particular, avoid a repetition of what is syntactically visible (takes a Widget and returns a boolean) and avoid exposing implementation details (run on every iteration).

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1817,22 +1854,31 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
 			Object data = widget.getData();
 			if (widget instanceof Item && data != null
-					&& !isExpandable((Item) widget, null, data)) {
+							&& !isExpandable((Item) widget, null, data)) {
```
Please revert formatting change.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1842,16 +1888,33 @@ protected void internalExpandToLevel(Widget widget, int level) {
 				if (children != null) {
-					int newLevel = (level == ALL_LEVELS ? ALL_LEVELS
-							: level - 1);
+					int newLevel = (level == ALL_LEVELS ? ALL_LEVELS : level - 1);
```
Please revert formatting change.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1817,22 +1854,31 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
+	 *                             further
+	 * @since 3.32
+	 */
+	protected void internalConditionalExpandToLevel(Widget widget, int level,
```
Can this method be private?

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2518,59 @@ public void setAutoExpandLevel(int level) {
+	 * expanded.
+	 * <p>
+	 * Auto Expansion of paths of consecutive single children is off by default.
+	 * Turning this behaviour off should be done cautiously on trees with
```
```suggestion
	 * Turning this behaviour on should be done cautiously on trees with
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2518,59 @@ public void setAutoExpandLevel(int level) {
+	 * lazy-loaded child-nodes.
+	 * <p>
+	 * <p>
+	 * Using <code> ALL_LEVELS </code> as arguments will set to infinite
```
```suggestion
	 * Using <code> ALL_LEVELS </code> as argument will set to infinite
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2518,59 @@ public void setAutoExpandLevel(int level) {
+	 * </p>
+	 *
+	 * @param level <code> NO_EXTEND </code> for disabled, <code> ALL_LEVELS </code>
+	 *              for infinite expansion or any integer value to set a level to
```
```suggestion
	 *              for infinite expansion or any positive integer value to set a level to
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -3426,4 +3543,18 @@ ISelection getUpdatedSelection(ISelection selection) {
 	}
 
+	/**
+	 * Returns whether the given Widget is a single child. Used as functional
```
```suggestion
	 * Returns whether the given Widget has exactly one child. Used as functional
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -3426,4 +3543,18 @@ ISelection getUpdatedSelection(ISelection selection) {
+	 * Returns whether the given Widget is a single child. Used as functional
+	 * Parameter to {@link #internalConditionalExpandToLevel(Widget, int, Function)}
+	 * while expanding paths of single children
+	 * {@link #setAutoExpandOnSingleChildLevel(int)}
```
Please do not document the users of a method in the method documentation.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -3426,4 +3543,18 @@ ISelection getUpdatedSelection(ISelection selection) {
+	 * while expanding paths of single children
+	 * {@link #setAutoExpandOnSingleChildLevel(int)}
+	 *
+	 * @param Widget
```
Parameter documentation is missing. Actually, I would remove the Javadoc for this method, as it is a completely self-explaining, private method.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2518,59 @@ public void setAutoExpandLevel(int level) {
+	}
+
+	private void removeAutoExpandOnSingleChildListener() {
+		if (autoExpandOnSingleChildListener != null) {
```
You can avoid this null check, as the called method properly handles null anyway. But you should assign `null` to `autoExpandOnSingleChildListener` in this method.

Wittmaxi replied:
The [method that removes the listener from the list](https://github.com/eclipse-equinox/equinox/blob/66972dd6c7b1e63106505b3cd67d0a034b4c817f/bundles/org.eclipse.equinox.common/src/org/eclipse/core/runtime/ListenerList.java#L221) will throw an `IllegalArgumentException` if the listener is null.
If we want to remove the `null`-check, we have a few options:
- Allow the method throw in case the listener is null and make the user handle this: I really don't like this option since this exposes internal state to the user. A trivial way to reach this state is eg. if the user calls `setAutoExpandOnSingleChildListener(NO_EXPAND)` before anything else - as done in our own unit test [here](https://github.com/eclipse-platform/eclipse.platform.ui/pull/1072/files#diff-82159efdd73bf97fd8fdde7cb0215f0fd2e302f1c87c2c0423d28e644c230837R190)
- Catch the exception ourselves: IMHO, this obfuscates meaning more than it helps.

Since I deem both options unsatisfactory, I have re-enabled the null-check.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2518,59 @@ public void setAutoExpandLevel(int level) {
 	}
 
+	/**
+	 * Sets the level of auto expand on paths of consecutive single children.
```
I find this description hardly comprehensible since you rely on intuition for understanding what a "path of consecutive single children" is. What about something like: "Sets the maximum number of levels to automatically expand whenever a widget is expanded that only has a single child element. The expansion is recursively applied for the given number of levels if the widget at each of the according levels only has a single child."

HeikoKlare replied:
What information then still is missing is _when_ this mechanism is applied? It cannot be whenever a node is expanded, as the functionality would then apply recursively. It can also not be whenever a node is not expanded by this mechanism, as, e.g., the `expandToLevel` functionality expands nodes in a way that this new auto-expand functionality is not applied. The documentation should precisely state when the expansion happens. Right now, this is whenever an `SWT.EXPAND` event occurs on the underlying tree and whenever the `setExpandedState` method is called (which in my opinion is not correct, see according comment).

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2518,59 @@ public void setAutoExpandLevel(int level) {
+	/**
+	 * Sets the level of auto expand on paths of consecutive single children.
+	 * <code> NO_EXPAND </code> means that such paths are not automatically
+	 * expanded.
```
I'd like to have a more precise description of the meaning of the given number of levels. E.g., does `level=1` mean that one _additional_ level is expanded in case an item with one child is expanded or that the the one level that is expanded anyway is expanded? Currently it seems to be latter one, i.e., `level=0` and `level=1` will have the same behavior. That behavior is fine for me, but should be precisely documented.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -102,8 +112,24 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
+	 * will disable the feature, while a negative value (<code>ALL_LEVELS</code>)
+	 * will expand to infinity
+	 */
+	private int autoExpandOnSingleChildLevel = 0;
```
Should this be called `autoExpandOnSingleChildLevels`?` It does not indicate a level but a number of levels. Same applied to the according method.

Wittmaxi replied:
Since this behaves like `setAutoExpandLevel` with a special condition, I don't see what warrants the inconcistency in plural/singular. "Level" refers to how deep a recursive path with singular descendants will be expanded - what is the meaning behind talking about "multiple levels"?

HeikoKlare replied:
I mentioned this, because the behavior is actually different from `expandToLevel` / `setAutoExpandLevel`. That existing expand level variable defines the level to which auto-expansion is performed. It does not define a number of (relative) levels to expand but defines precisely one level, to whcih expansion is done. The `autoExpandOnSingleChildLevel`, however, does not define a specific level, to which the expansion is performed, but defines _a number of levels_ to expand.

Wittmaxi replied:
Now I understand your comment, I have addressed it in the newest forcepush

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2573,24 +2689,25 @@ public int hashCode(Object element) {
 	 * Sets whether the node corresponding to the given element or tree path is
-	 * expanded or collapsed.
+	 * expanded or collapsed. Will auto expand paths of single-children if setup by
+	 * {@link #setAutoExpandOnSingleChildLevel(int) }
```
This is a breaking API change. It modifies the postcondition of a public API method, thus changes its behavior and requires a potential adaptation of all implementations (i.e., sublasses overwriting the method).
The method is, e.g., also used to perform refresh operations, which first collapse the element and then expand it again. It is unexpected for these callers (and the users of those refresh operations) that the tree items suddenty expand on refresh operations if the new auto-expand functionality is enabled.

I think we need to add a new method like `setExpandedStateWithAutoExpandOnSingleChild` providing that functionality. We then have to check the existing callers of `setExpandedState` for whether they can be should be changed to calls of the new methods. It becomes clear that adaptations of existing users may be necessary when the documentation of `setAutoExpandOnSingleChildLevel` properly states when the auto-expand mechanism will be used.

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -151,6 +151,53 @@ public void testExpandToLevel() {
+		assertTrue("The second child of the trivial path was not auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild().getFirstChild()));
+		assertTrue("The third child of the trivial path was not auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild().getFirstChild().getFirstChild()));
```
It would be good to have a loop over all childs checking that they are expanded rather than only checking an arbitrary number of levels.

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -151,6 +151,53 @@ public void testExpandToLevel() {
+		TestElement trivialPathRoot = modelRoot.getFirstChild();
+		fViewer.setInput(modelRoot);
+
+		fTreeViewer.setAutoExpandOnSingleChildLevel(AbstractTreeViewer.NO_EXPAND);
```
Since this is the default state, it would also be interesting to have a state for first enabling and then re-disabling the auto-expansion mechanism.

HeikoKlare reviewed (changes requested):
I have two major remarks (and requests for change) and some minor detailed comments:
1. The current changes break public API (`setExpandedState`). This needs to be adapted.
2. The tests should cover all different usage scenarios of the different methods. In particular, they should cover the different ways in which auto-expansion can happen, i.e., not only via `setExpandedState` but also via SWT events. I'd propose to relate the tests to the different added or changed methods. For example, there should also be trivial tests, such as `get...Level` returning what has been set by `set...Level` to easy debugging in case only the setter is broken, as currently the tests for the expand behavior would break although it's state may simply not be represented as expected after calling the setter.

Wittmaxi commented:
The `continuous-integration/jenkins/pr-head` build error stems from the org.osgi.http - package, unrelated to my PR.
All unit tests could be built

iloveeclipse commented:
> The `continuous-integration/jenkins/pr-head` build error stems from the org.osgi.http - package, unrelated to my PR.

Please **rebase** on master.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -83,27 +84,52 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
-	 * of 0 means that auto-expand is off.
+	 * The level to which the tree is automatically expanded each time the viewer's
+	 * input is changed (that is, by <code>setInput</code>). A value of {code
+	 * NO_EXPAND } means that auto-expand is off.
```
```suggestion
	 * input is changed (that is, by <code>setInput</code>). A value of @{code
	 * NO_EXPAND} means that auto-expand is off.
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -83,27 +84,52 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
 	/**
-	 * Indicates if filters should be checked to determine expandability of
-	 * a tree node.
+	 * How many levels to autoexpand in a path of consecutive single children. {code
```
```suggestion
	 * How many levels to autoexpand in a path of consecutive single children. @{code
```

HeikoKlare replied:
Please use a better description than "consecutive single children", see https://github.com/eclipse-platform/eclipse.platform.ui/pull/1072#discussion_r1356711230.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -83,27 +84,52 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
-	 * Indicates if filters should be checked to determine expandability of
-	 * a tree node.
+	 * How many levels to autoexpand in a path of consecutive single children. {code
+	 * NO_EXPAND} will disable the feature, while <code>ALL_LEVELS</code> will
```
Is there a reason for sometimes using `{@code ...}` and other times using `<code>...</code>`?

Wittmaxi replied:
This inconsistency is not intended. From [this documentation](https://docs.oracle.com/javase/8/docs/technotes/tools/unix/javadoc.html#CHDJGIJB), I infer that `{@code ...}` is slightly better fit for our usecase. I will check through the PR to make sure that I consistently use the correct JavaDoc-Tag - unless otherwise warranted

HeikoKlare replied:
That's conform to my understanding. Making this consistent would be good. I was just wondering if there is a specific reason for the inconsistent way of formatting (because of interest and because others may ask theirselves the same when seeing such inconsistencies).

Wittmaxi replied:
IIRC, I used `<code> ... </code>` because the rest of the file used it a lot. I don't believe that refactoring the entire file for this consistency is a project for this PR - We could however choose to create a PR where we change all occurences of `<code> ... </code>` to `{@code ...}`. If you agree with me, I will open an issue to put this up for discussion

HeikoKlare replied:
No, I did not want to propose to cleanup the entire file. I just wanted to have consistency within the text that this PR touches. And there is no need to change the rest of the documentation or open an issue for that.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -83,27 +84,52 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
-	 * a tree node.
+	 * How many levels to autoexpand in a path of consecutive single children. {code
+	 * NO_EXPAND} will disable the feature, while <code>ALL_LEVELS</code> will
+	 * expand to infinity
```
```suggestion
	 * expand to the leaves.
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -83,27 +84,52 @@ public abstract class AbstractTreeViewer extends ColumnViewer {
+
+	/**
+	 * Listens to expansion events and recursively expands all children which
+	 * themselves only contain one single child.
```
This description is not correct, as the listener does not expand _all_ children. You can keep this more simple by saying that this `Listens to expansion events for triggering the auto expansion mechanism for elements with a single child according to {@link #setAutoExpandOnSingleChildLevels(int)}.` or the like

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1221,6 +1247,17 @@ public int getAutoExpandLevel() {
+	/**
+	 * @return <code> NO_EXPAND </code> for disabled, <code> ALL_LEVELS </code> for
+	 *         infinite expansion or any integer value for the currently set level
+	 *         of expansion
```
```suggestion
	 * @return <code>NO_EXPAND</code> for disabled, <code>ALL_LEVELS</code> for
	 *         infinite expansion or any integer value for a specific number of levels to expand
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1817,18 +1854,23 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
-	 *               levels of the tree
+	 * @param widget               the widget
+	 * @param level                non-negative level, or <code>ALL_LEVELS</code> to
+	 *                             collapse all levels of the tree
```
This is about expand and not collapse, is it? I know the documentation was like this before, but probably it was incorrect?
```suggestion
	 *                             expand all levels of the tree
```

Wittmaxi replied:
The expansion or collapsing of the widget depends on the third parameter to this method;
```java
if (level == ALL_LEVELS || level > 1) {
Item[] children = getChildren(widget);
if (children != null) {
	int newLevel = (level == ALL_LEVELS ? ALL_LEVELS
			: level - 1);
	for (Item element : children) {
		if (shouldChildrenExpand.apply(widget).booleanValue()) {
			internalConditionalExpandToLevel(element, newLevel, shouldChildrenExpand);
		}
	}
}
}
```
As such, neither words are really well-fit to fully describe the behaviour. What about something like "non-negative level, or {@code ALL_LEVELS} to which paths of a tree should either be collapsed or expanded"?

HeikoKlare replied:
From my understanding, the description is fine as already proposed. My concern was only regarding "collapsed" as the method only seems to expand. Is there really functionality to collapse inside this method?

Wittmaxi replied:
I have double-checked, you are indeed right, there is no code to collapse widgets in the method, hence `expand` is the right terminology

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1845,13 +1887,31 @@ protected void internalExpandToLevel(Widget widget, int level) {
+	 * </p>
+	 *
+	 * @param widget the widget
+	 * @param level  non-negative level, or <code>ALL_LEVELS</code> to collapse all
```
see above
```suggestion
	 * @param level  non-negative level, or <code>ALL_LEVELS</code> to expand all
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * <p>
+	 * This behaviour applies when expanding a Widget using the method
+	 * {@link #setExpandedStateWithAutoExpandOnSingleChild(Object, boolean)} or when
+	 * an <code> SWT.expand </code> event is triggered on the underlying tree. The
```
```suggestion
	 * an {@link SWT#Expand} event is triggered on the underlying tree. The
```
or
```suggestion
	 * an <code>SWT.Expand</code> event is triggered on the underlying tree. The
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * {@link #setExpandedStateWithAutoExpandOnSingleChild(Object, boolean)} or when
+	 * an <code> SWT.expand </code> event is triggered on the underlying tree. The
+	 * behaviour is such, that a level of 0 or 1 represents opening the current
+	 * Widget itself. Thus, only a <code>level </code> of 2 or greater will have any
```
```suggestion
	 * Widget itself. Thus, only a <code>level</code> of 2 or greater will have any
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * This behaviour applies when expanding a Widget using the method
+	 * {@link #setExpandedStateWithAutoExpandOnSingleChild(Object, boolean)} or when
+	 * an <code> SWT.expand </code> event is triggered on the underlying tree. The
+	 * behaviour is such, that a level of 0 or 1 represents opening the current
```
```suggestion
	 * behavior is such, that a level of 0 or 1 represents opening the current
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * automatically expanded.
+	 * <p>
+	 * Auto Expansion of paths of consecutive single children is off by default.
+	 * Turning this behaviour on should be done cautiously on trees with lazy-loaded
```
```suggestion
	 * Turning this behavior on should be done cautiously on trees with lazy-loaded
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * an <code> SWT.expand </code> event is triggered on the underlying tree. The
+	 * behaviour is such, that a level of 0 or 1 represents opening the current
+	 * Widget itself. Thus, only a <code>level </code> of 2 or greater will have any
+	 * additional effect. <code> NO_EXPAND </code> means that such paths are not
```
```suggestion
	 * additional effect. <code>NO_EXPAND</code> means that such paths are not
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * additional effect. <code> NO_EXPAND </code> means that such paths are not
+	 * automatically expanded.
+	 * <p>
+	 * Auto Expansion of paths of consecutive single children is off by default.
```
Please correct old terminoloty "path of consecutive single children".

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * <p>
+	 * <p>
+	 * Using <code> ALL_LEVELS </code> as argument will set to infinite
+	 * auto-expansion of paths of consecutive single children.
```
Please correct old terminoloty "path of consecutive single children".

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * child-nodes.
+	 * <p>
+	 * <p>
+	 * Using <code> ALL_LEVELS </code> as argument will set to infinite
```
```suggestion
	 * Using <code>ALL_LEVELS</code> as argument will set to infinite
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 * auto-expansion of paths of consecutive single children.
+	 * </p>
+	 *
+	 * @param level <code> NO_EXPAND </code> for disabled, <code> ALL_LEVELS </code>
```
```suggestion
	 * @param level <code>NO_EXPAND /code> for disabled, <code>ALL_LEVELS</code>
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2515,69 @@ public void setAutoExpandLevel(int level) {
+	 *
+	 * @param level <code> NO_EXPAND </code> for disabled, <code> ALL_LEVELS </code>
+	 *              for infinite expansion or any positive integer value to set a
+	 *              level to which paths of consecutive single children are
```
Please correct old terminoloty "path of consecutive single children".

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2594,6 +2715,29 @@ public void setExpandedState(Object elementOrTreePath, boolean expanded) {
+	/**
+	 * Behaves like {{@link #setExpandedState(Object, boolean)} but additionally
+	 * also respects expansion of paths of single-children if setup by
+	 * {@link #setAutoExpandOnSingleChildLevels(int)}
```
```suggestion
	 * Behaves like {@link #setExpandedState(Object, boolean} but additionally
	 * also respects expansion of single child elements if set up by
	 * {@link #setAutoExpandOnSingleChildLevels(int)}.
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2594,6 +2715,29 @@ public void setExpandedState(Object elementOrTreePath, boolean expanded) {
+	 * {@link #setAutoExpandOnSingleChildLevels(int)}
+	 *
+	 * @param elementOrTreePath
+	 * @param expanded
```
Documentation is missing

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2594,6 +2715,29 @@ public void setExpandedState(Object elementOrTreePath, boolean expanded) {
+			return;
+		}
+		setExpandedState(elementOrTreePath, expanded);
+		Widget item = internalExpand(elementOrTreePath, false);
```
This does not seem to be necessary, as the element has already been expanded in the line below. It seems to be a hack to retrieve the element at the given path. This can also be achieved by calling `internalGetWidgetToSelect()`.

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,87 @@ public void testExpandToLevel() {
+
+		assertTrue("The expanded widget child is not expanded", fTreeViewer.getExpandedState(trivialPathRoot));
+		assertTrue("The first child of the trivial path was auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()) == false);
```
```suggestion
		assertFalse("The first child of the trivial path was auto-expanded",
				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()));
```

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,87 @@ public void testExpandToLevel() {
+		assertTrue("The first child of the trivial path was not auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()));
+		assertTrue("Trivial path is expanded further than specified depth ",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild().getFirstChild()) == false);
```
```suggestion
		assertFalse("Trivial path is expanded further than specified depth",
				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild().getFirstChild()));
```

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,87 @@ public void testExpandToLevel() {
+
+		assertTrue("The expanded widget child is not expanded", fTreeViewer.getExpandedState(trivialPathRoot));
+		assertTrue("The first child of the trivial path was auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()) == false);
```
```suggestion
		assertFalse("The first child of the trivial path was auto-expanded",
				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()));
```

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,87 @@ public void testExpandToLevel() {
+		TestElement child = trivialPathRoot.getFirstChild();
+		int depth = 0;
+		while (child != null) {
+			assertTrue("The " + ++depth + ". child of the trivial path was not auto-expanded",
```
To avoid fixed inline incrementation and string generation, you could also do:
```suggestion
for (int depth = 1; child != null; depth++) {
			assertTrue("The " + depth + ". child of the trivial path was not auto-expanded",
```

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/TreeViewerTest.java:
```diff
@@ -50,4 +55,22 @@ protected String getItemText(int at) {
+		assertTrue("The first child of the trivial path was not auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()));
+		assertTrue("Trivial path is expanded further than specified depth ",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild().getFirstChild()) == false);
```
```suggestion
		assertFalse("Trivial path is expanded further than specified depth",
				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild().getFirstChild()));
```

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,87 @@ public void testExpandToLevel() {
 	}
 
 	@Test
+	public void testAutoExpandTrivialPath() {
```
It would be good to have two additional test cases:
1. A test that does not expand the root element but some child and is then asserted to expand the proper number of elements.
2. A test in which an element is expanded whose child depth is lower than the `autoExpandOnSingleChildLevels` value.

HeikoKlare reviewed (commented):
From a functional perspective, the code seems to be good now. I have tested the `CommonViewer` implementation of the `ProjectExplorer`, which appears to work fine with the new functionality.

The things lefts to improve are minor and they particularly concern documentation.

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2594,6 +2718,29 @@ public void setExpandedState(Object elementOrTreePath, boolean expanded) {
+	 * respects expansion of paths of of single child elements if set up by
+	 * {@link #setAutoExpandOnSingleChildLevels(int)}.
+	 *
+	 * @param elementOrTreePath The Widget to expand as described above
```
```suggestion
	 * @param elementOrTreePath the widget to expand
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2594,6 +2718,29 @@ public void setExpandedState(Object elementOrTreePath, boolean expanded) {
+	 * {@link #setAutoExpandOnSingleChildLevels(int)}.
+	 *
+	 * @param elementOrTreePath The Widget to expand as described above
+	 * @param expanded          The new state of {@code elementOrTreePath}
```
```suggestion
	 * @param expanded          the new expanded state of {@code elementOrTreePath}
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1817,18 +1855,23 @@ protected Widget internalGetWidgetToSelect(Object elementOrTreePath) {
+	 * @param widget               the widget
+	 * @param level                non-negative level, or {@code ALL_LEVELS} to
+	 *                             expand all levels of the tree
+	 * @param shouldChildrenExpand predicate that defines for a given Widget should
```
```suggestion
	 * @param shouldChildrenExpand predicate that defines for a given widget if it should
```

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -1221,6 +1248,17 @@ public int getAutoExpandLevel() {
+	/**
+	 * @return {@code NO_EXPAND} for disabled, {@code ALL_LEVELS} for infinite
+	 *         expansion or any integer value for the currently set level of
+	 *         expansion
```
See https://github.com/eclipse-platform/eclipse.platform.ui/pull/1072#discussion_r1399131376

Wittmaxi replied:
https://github.com/eclipse-platform/eclipse.platform.ui/pull/1072/files#diff-71134aa69d7c8b54cb0e0f2ed6069aa3ee529413112c03a3127cad7101b517e2R1251-R1257

Already reads ``` * @return {@code NO_EXPAND} for disabled, {@code ALL_LEVELS} for infinite
	 *         expansion or any integer value for a specific number of levels to
	 *         expand.```, as expected by that review-comment. What am I misunderstanding?

HeikoKlare replied:
No misunderstanding. It's only that you pushed the fix 20 minutes after I made the comment 🙂 
https://github.com/eclipse-platform/eclipse.platform.ui/commit/b8234bec3634e0892dfa71f44cff1a5694d2027c -> https://github.com/eclipse-platform/eclipse.platform.ui/commit/329a08848ddce7b678d9e1cfc7177e6a47370341

HeikoKlare commented on bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java:
```diff
@@ -2455,6 +2516,71 @@ public void setAutoExpandLevel(int level) {
+	 * <p>
+	 * This behavior applies when expanding a Widget using the method
+	 * {@link #setExpandedStateWithAutoExpandOnSingleChild(Object, boolean)} or when
+	 * an {@code SWT.expand} event is triggered on the underlying tree. The behavior
```
See https://github.com/eclipse-platform/eclipse.platform.ui/pull/1072#discussion_r1399394845

HeikoKlare reviewed (commented):
Some remaining documentation adaptations noted in the detailed comments.
Please also consider https://github.com/eclipse-platform/eclipse.platform.ui/pull/1072#discussion_r1399115018.

HeikoKlare commented on bundles/org.eclipse.jface/.settings/.api_filters:
```diff
@@ -42,4 +42,12 @@
             </message_arguments>
         </filter>
     </resource>
+    <resource path="src/org/eclipse/jface/viewers/AbstractTreeViewer.java" type="org.eclipse.jface.viewers.AbstractTreeViewer">
+        <filter id="336658481">
+            <message_arguments>
+                <message_argument value="org.eclipse.jface.viewers.AbstractTreeViewer"/>
+                <message_argument value="NO_EXPAND"/>
+            </message_arguments>
+        </filter>
+    </resource>
```
We need minor version bump for the jface plugin because of the API extension. From my understanding, this API filter should be obsolete when adding that version bump.

Wittmaxi replied:
Fixed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/4f7cd31362a686529d68f8ef6d0c33e2c779403a

HeikoKlare reviewed (commented):
Thank you! Changes look good now. We need to retrigger the build once master is open again and #1321 has been merged to see if builds run fine.

Just one minor thing: the version bump commit now contains changes not related to the version but to the feature itself (such as the API filters addition). The changes of the `@since` tags in that commit are somehow valid (because the commit bumps the version accordingly) but the feature cannot be added without the version bump anyway, so they might also belong to the feature commit as well. Maybe it still makes sense to combine both commits, as the version bump is tightly related to the feature (as that's what requires a minor rather than a micro version bump)?

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,144 @@ public void testExpandToLevel() {
+		TestElement modelRoot = TestElement.createModel(6, 1);
+		TestElement trivialPathRoot = modelRoot.getFirstChild();
+		fViewer.setInput(modelRoot);
+
```
This test fails on Linux because it does not automatically expand parent nodes and the root node is unexpanded here. Since that is OS-dependent behavior and nothing to be tested here, you can simply expand the root note at the beginning of the test and proceed as is.
`fTreeViewer.setExpandedState(trivialPathRoot, true);`

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/TreeViewerTest.java:
```diff
@@ -50,4 +56,25 @@ protected String getItemText(int at) {
+		fTreeViewer.setAutoExpandOnSingleChildLevels(2);
+		Event event = new Event();
+		System.out.println("TreeViewerTest - 1: " + event.item);
+		event.item = fViewer.testFindItem(trivialPathRoot);
```
Can you please try replacing this line as follows:
```suggestion
		event.item = ((Tree) fViewer.getControl()).getItem(0);
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/TreeViewerTest.java:
```diff
@@ -50,4 +56,25 @@ protected String getItemText(int at) {
 	}
 
+	@Test
+	public void testAutoExpandTrivialPathThroughEvent() {
```
```suggestion
	public void testAutoExpandOnSingleChildThroughEvent() {
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,144 @@ public void testExpandToLevel() {
+	}
+
+	@Test
+	public void testAutoExpandTrivialPathManualDisable() {
```
```suggestion
	public void testAutoExpandOnSingleChildManualDisable() {
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,144 @@ public void testExpandToLevel() {
+	}
+
+	@Test
+	public void testAutoExpandTrivialPathSmallerThanAutoExpandDepth() {
```
```suggestion
	public void testAutoExpandOnSingleChildSmallerThanAutoExpandDepth() {
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,144 @@ public void testExpandToLevel() {
+	}
+
+	@Test
+	public void testAutoExpandTrivialPathFromRootWithSensibleTrivialPath() {
```
```suggestion
	public void testAutoExpandOnSingleChildFromRootWithSensibleTrivialPath() {
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,144 @@ public void testExpandToLevel() {
+	}
+
+	@Test
+	public void testAutoExpandTrivialPathFromRoot() {
```
```suggestion
	public void testAutoExpandOnSingleChildFromRoot() {
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,144 @@ public void testExpandToLevel() {
+	}
+
+	@Test
+	public void testAutoExpandTrivialDeeperDownPath() {
```
```suggestion
	public void testAutoExpandOnSingleChildDeeperDownPath() {
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java:
```diff
@@ -164,6 +164,144 @@ public void testExpandToLevel() {
 	}
 
 	@Test
+	public void testAutoExpandTrivialPath() {
```
```suggestion
	public void testAutoExpandOnSingleChild() {
```

Wittmaxi replied:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/fb256ed9423ce9f552aa278fa3893d833ccc0e63

HeikoKlare reviewed (commented):
The test `testAutoExpandOnSingleChildDeeperDownPath` is now always failing, as the expand state of the root node is set before the input is actually set, so this will have no effect. Please reorder the statements. You should be able to test this locally. 

Please also remove the debug output from the tests again and rebase on master. Currently, there is a conflict in the `.api_filters` file.

Commit:
TreeViewer: AutoExpandTrivialPaths

This commit addresses the need to automatically expand paths which
consist of chained singular nodes, eg:

com.wittmaxi.plugin
 └─src
    └─org.foo.com
       └─Bar.java

It introduces two new methods to the API of TreeViewer

-setAutoExpandOnSingleChildLevel
-getAutoExpandOnSingleChildLevel

Also includes a version bump for AbstractTreeViewer Autoexpand

HeikoKlare committed:
Disable auto-expand on single child test for lazy tree viewer on Mac

Tests for virtual lazy tree viewers that access the data of a tree items
have been disabled because of some bug in according Mac implementation.
This also disables a test case for the auto-expand on single child
functionality that relies on accessing the item data.

Also removes unused API problem filters.

HeikoKlare reviewed (approved):
Changes work and look good now.

I propose to target the merge for 2024-06 M1. The change affects the `TreeViewer` as a rather central component, so I would propose to not risk a potentially undetected regression now that we are already quite late in 2024-03 development.

HeikoKlare committed:
Version bump for 4.32 stream

HeikoKlare committed:
Remove duplicate and obsolete paragraph tags in Javadoc

HeikoKlare commented:
Failing test is pre-existing and already documented: #1005.

HeikoKlare reviewed (commented):
This turned out to contain a kind of breaking API change:
The method `protected void internalExpandToLevel(Widget widget, int level)` of `AbstractTreeViewer` used to call itself recursively before the change, while after the change it just calls a separate method. When some subclass had overwritten that method intercepting the recursive call (e.g., for measuring performance of subtree expansions), this is not working anymore, i.e., something like:
```java
public class MyTree extends AbstractTreeViewer {
   ...
   @Override
   protected void internalExpandToLevel(Widget widget, int level) {
      // start measuring
      super.internalConditionalExpandToLevel(widget, level);
      // stop measuring
   }
````

I found nothing about this specific topic (changing recursive behavior of public/protected method) in the documentation on evolving APIs (https://github.com/eclipse-platform/eclipse.platform/blob/master/docs/Evolving-Java-based-APIs.md etc.), but in my opinion it can result in something appearing like a breaking API change to consumers. @akurtakov do you have any opinion or insights on this kind of issue?
